src/codeGenerator/evaluator.ts

```diff
@@ -0,0 +1,333 @@
+        }
+    }
+
+    visitImport(_astNode: Import, _env: EvaluatorEnv) : EvaluatorData {
+        throw new Error("Method not implemented.");
+    }
+    visitDefine(astNode: Define, env: EvaluatorEnv) : EvaluatorData {
+        let fnArgNames = new Set();
+        for (let fnArgName of astNode.args) {
+            if (fnArgNames.has(fnArgName.id)) {
+                throw new DSLSyntaxError(astNode, `define: Duplicate function argument name ${fnArgName}`);
+            }
+            fnArgNames.add(fnArgName.id);
+        }
+
+        let closure = new FunctionClosure(astNode, env);
+        if (astNode.id !== null) {
+            // IDs at binding time are treated as raw strings, and not evaluated further
+            env.extend(astNode.id.id, closure, true);
+        }
+        return closure;
+    }
+    visitLet(astNode: Let, env: EvaluatorEnv) : EvaluatorData {
+        if (astNode.ids.length !== astNode.values.length) {
+            throw new DSLSyntaxError(astNode,
+                `let: Lengths of IDs (${astNode.ids.length}) and expressions (${astNode.values.length}) do not match`);
+        }
+        let newEnv = env;
+        for (let idx in astNode.ids) {
+            let id = astNode.ids[idx];
+            if (id == null) {
+                throw new DSLSyntaxError(astNode, `let: Expected string ID in index ${idx}, got ${id}`);
+            }
+
+            let expr = astNode.values[idx];
+            if (expr == null) {
+                throw new DSLSyntaxError(astNode, `let: Expected expression in index ${idx}, got ${expr}`);
+            }
+
+            let result = expr.accept(this, env);
+            // IDs at binding time are treated as raw strings, and not evaluated further
+            newEnv = newEnv.extend(id.id, result);
+        }
+        return astNode.body.accept(this, newEnv);
+    }
+    visitIf(astNode: If, env: EvaluatorEnv) : EvaluatorData {
+        if (astNode.predicate.accept(this, env)) {
+            return astNode.consequent.accept(this, env);
+        } else {
+            return astNode.alternative.accept(this, env);
+        }
+    }
+    visitFor(astNode: For, env: EvaluatorEnv) : EvaluatorData {
+        let forResult: EvaluatorData[] = [];
+        let variableName = astNode.id.id;
+        let iterableValue = astNode.iterable.accept(this, env);
+        if (!Array.isArray(iterableValue)) {
+            throw new DSLTypeError(astNode, `for: expected list as target, got type ${typeof iterableValue}`)
+        }
+        for (let elem of iterableValue) {
+            let newEnv = env.extend(variableName, elem);
+            let result = astNode.body.accept(this, newEnv);
+            forResult.push(result);
+        }
+        return forResult;
+    }
+    visitPrint(astNode: Print, env: EvaluatorEnv) : EvaluatorData {
+        let s = astNode.expression.accept(this, env);
+        console.log(s);
+        return s;
+    }
+    visitBinop(astNode: Binop, env: EvaluatorEnv) : EvaluatorData {
```

Comment:
Ok, we should define how we want the operators to behave - otherwise we'll just fall back to JavaScript's hall of messy operations.

Does this seem reasonable:

- `&&` and `||` can probably work on all types and check whether they're truthy? This is what `if` does already
- Add on two strings concats two strings
- Adding a number to a string concats the number to the string (since we don't have explicit casting)?
- Adding a string to a number is ambiguous and should be rejected
- All other arithmetic fns only accept numbers
- `>`/`>=`/`<`/`<=` can compare numbers against numbers and strings against strings